lib/routes/recovery-keys.js

```diff
@@ -0,0 +1,230 @@
+          .then(createRecoveryKey)
+          .then(emitMetrics)
+          .then(() => reply({}), reply)
+
+        function createRecoveryKey() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.createRecoveryKey(uid, recoveryKeyId, recoveryData)
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'account.recoveryKey.created',
+            uid: uid
+          })
+
+          return request.emitMetricsEvent('recoveryKey.created', {uid: uid})
+            .then(() => ({}))
+        }
+      }
+    },
+    {
+      method: 'GET',
+      path: '/recoveryKeys/{recoveryKeyId}',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        }
+      },
+      handler(request, reply) {
+        log.begin('getRecoveryKey', request)
+
+        const uid = request.auth.credentials.uid
+        const sessionToken = request.auth.credentials
+        const recoveryKeyId = request.params.recoveryKeyId
+        let recoveryData
+
+        customs.check(request, 'getRecoveryKey')
+          .then(getRecoveryKey)
+          .then(() => reply({recoveryData}), reply)
+
+        function getRecoveryKey() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.getRecoveryKey(uid, recoveryKeyId)
+            .then((res) => recoveryData = res.recoveryData)
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/account/reset/recoveryKeys',
+      config: {
+        auth: {
+          strategy: 'accountResetToken'
+        },
+        validate: {
+          query: {
+            keys: isA.boolean().optional()
+          },
+          payload: {
+            authPW: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            wrapKb: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            recoveryKeyId: isA.string().max(256).required(),
+          }
+        }
+      },
+      handler: function (request, reply) {
+        log.begin('Password.changeRecoveryKeys', request)
+        const authPW = request.payload.authPW
+        const wrapKb = request.payload.wrapKb
+        const recoveryKeyId = request.payload.recoveryKeyId
+        const uid = request.auth.credentials.uid
+
+        let authSalt, password, verifyHash, sessionToken, keyFetchToken, account
+
+        checkRecoveryKey()
+          .then(getAccount)
+          .then(passwordReset)
+          .then(deleteRecoveryKey)
+          .then(notifyAccount)
+          .then(createSessionToken)
+          .then(createKeyFetchToken)
+          .then(createResponse)
+          .then(reply, reply)
+
+        function checkRecoveryKey() {
+          return db.getRecoveryKey(uid, recoveryKeyId)
+        }
+
+        function getAccount() {
+          return db.account(uid)
+            .then((result) => account = result)
+        }
+
+        function passwordReset() {
```

Comment:
I feel like I don't understand how the parts fit together here, does this work like a password change where we use the new password to wrap the original wrapKb so the user can still decrypt their data encrypted using the original kB?

Reply:
Sorry I haven't had a chance to go through and add comments yet. You are correct though, this will basically be a password change using the `wrapKb` containing the original kB.

I think the best summary of registering steps involved are [here](https://docs.google.com/document/d/1LrIblRRHIuEQwGNwKYY29M7LWO-a8ng-qnR8tNJAg8U/edit#heading=h.nq5k8b9xpezv) and consuming steps [here](https://docs.google.com/document/d/1LrIblRRHIuEQwGNwKYY29M7LWO-a8ng-qnR8tNJAg8U/edit#heading=h.vhp1wq6l80f7).